### PR TITLE
fix: sensitive data compare should use constant time compare to avoid timing attack

### PR DESCRIPTION
**Pull request by nan01ab**
# Description

`RpcServer.CheckAuth` use direct compare to determine two `string`s equal or not, but sensitive data compare should use constant time compare to avoid timing attack

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by vncoelho — COMMENTED** [review]
I understand this timing attack. It looks correct and the XOR is quite efficient.
Although it does not look much critical to me in this specific case.

**Review by vncoelho — DISMISSED** [review]

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
side-channel attack is sort of hort in the research~~~~ but not sure how industry see it.

File: src/Neo.Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -43,5 +43,30 @@ public static int GetVarSize(this string value)
...
+        public static bool ConstantTimeEquals(this string left, string right)
+        {
+            if (left == null && right == null)
+                return true;
+
+            if (left == null || right == null)
+                return false;
+
+            if (left.Length != right.Length)
+                return false;
```
Comment (shargon) [inline src/Neo.Extensions/StringExtensions.cs:62]: This is no constant, maybe we should do a dummy for for the maxLength string

File: src/Neo.Extensions/StringExtensions.cs [hunk 2]
```diff
@@ -43,5 +43,30 @@ public static int GetVarSize(this string value)
...
             return UnsafeData.GetVarSize(size) + size;
         }
+
+        /// <summary>
+        /// Compares two <see cref="string"/>s for equality in constant time.
+        /// </summary>
+        /// <param name="left">The left <see cref="string"/>.</param>
+        /// <param name="right">The right <see cref="string"/>.</param>
+        /// <returns>True if the two <see cref="string"/>s are equal, false otherwise.</returns>
+        public static bool ConstantTimeEquals(this string left, string right)
```
Comment (roman-khimov) [inline src/Neo.Extensions/StringExtensions.cs:53]: Can we please _not_ introduce yet another constant time comparison function? This was discussed already in https://github.com/neo-project/neo/pull/3472#pullrequestreview-2266994771, there are standard solutions to this that should be used.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: src/Neo.Extensions/StringExtensions.cs (hunk 2 above)
Comment (nan01ab) [reply src/Neo.Extensions/StringExtensions.cs:53]: > Can we please _not_ introduce yet another constant time comparison function? This was discussed already in [#3472 (review)](https://github.com/neo-project/neo/pull/3472#pullrequestreview-2266994771), there are standard solutions to this that should be used.

The implementation in `ConstantTimeUtility` has some limits.
For example:
`public static bool ConstantTimeEq<T>(in T a, in T b) **where T : unmanaged**`

File: src/Neo.Extensions/StringExtensions.cs (hunk 1 above)
Comment (nan01ab) [reply src/Neo.Extensions/StringExtensions.cs:62]: > This is no constant, maybe we should do a dummy for for the maxLength string

Implementations  in Rust `subtle` and Golang `crypto/subtle`  are similar

File: src/Neo.Extensions/StringExtensions.cs (hunk 2 above)
Comment (shargon) [reply src/Neo.Extensions/StringExtensions.cs:53]:
```csharp
        string string1 = "hello";
        string string2 = "hello";

        // Convert to bytes
        byte[] bytes1 = Encoding.UTF8.GetBytes(string1);
        byte[] bytes2 = Encoding.UTF8.GetBytes(string2);

        // Use FixedTimeEquals
        bool result = CryptographicOperations.FixedTimeEquals(bytes1, bytes2);
```

That's not working?

File: src/Neo.Extensions/StringExtensions.cs (hunk 1 above)
Comment (shargon) [reply src/Neo.Extensions/StringExtensions.cs:62]: > > This is no constant, maybe we should do a dummy for for the maxLength string
> 
> Implementations in Rust `subtle` and Golang `crypto/subtle` are similar to this one. Length is not sensitive.

Then this implementations are wrong 🙃

File: src/Neo.Extensions/StringExtensions.cs (hunk 2 above)
Comment (nan01ab) [reply src/Neo.Extensions/StringExtensions.cs:53]: > ```cs
>         string string1 = "hello";
>         string string2 = "hello";
> 
>         // Convert to bytes
>         byte[] bytes1 = Encoding.UTF8.GetBytes(string1);
>         byte[] bytes2 = Encoding.UTF8.GetBytes(string2);
> 
>         // Use FixedTimeEquals
>         bool result = CryptographicOperations.FixedTimeEquals(bytes1, bytes2);
> ```
> 
> That's not working?

This needs extra data copies?

Comment (shargon) [reply src/Neo.Extensions/StringExtensions.cs:53]: First rule of cryptography, don't write your own cryptography. I'm with @roman-khimov, better to use the official one

Comment (nan01ab) [reply src/Neo.Extensions/StringExtensions.cs]: > First rule of cryptography, don't write your own cryptography. I'm with @roman-khimov, better to use the official one

Has changed to `CryptographicOperations.FixedTimeEquals`, and also eliminated the two memory allocations(`Encoding.UTF8.GetString(Convert.FromBase64String(reqauth.Replace("Basic ", "").Trim()));` and  `authstring.Split`), so it should have better performance.

File: src/Plugins/RpcServer/RpcServer.cs [hunk 3]
```diff
@@ -65,21 +75,24 @@ internal bool CheckAuth(HttpContext context)
...
-            if (authvalues.Length < 2)
+            int colonIndex = Array.IndexOf(auths, (byte)':');
+            if (colonIndex == -1)
                 return false;
 
-            return authvalues[0] == settings.RpcUser && authvalues[1] == settings.RpcPass;
+            byte[] user = auths[..colonIndex];
+            byte[] pass = auths[(colonIndex + 1)..];
+
+            return CryptographicOperations.FixedTimeEquals(user, _rpcUser) && CryptographicOperations.FixedTimeEquals(pass, _rpcPass);
```
Comment (shargon) [inline src/Plugins/RpcServer/RpcServer.cs:95]:
```suggestion
            // Execute both checks always but both must be true
            return CryptographicOperations.FixedTimeEquals(user, _rpcUser) & CryptographicOperations.FixedTimeEquals(pass, _rpcPass);
```

Comment (shargon) [reply src/Plugins/RpcServer/RpcServer.cs:95]: changed && to &

File: src/Plugins/RpcServer/RpcServer.cs [hunk 4]
```diff
@@ -65,21 +75,25 @@ internal bool CheckAuth(HttpContext context)
...
-            if (authvalues.Length < 2)
+            int colonIndex = Array.IndexOf(auths, (byte)':');
+            if (colonIndex == -1)
                 return false;
 
-            return authvalues[0] == settings.RpcUser && authvalues[1] == settings.RpcPass;
+            byte[] user = auths[..colonIndex];
+            byte[] pass = auths[(colonIndex + 1)..];
+
+            // Execute both checks always but both must be true
```
Comment (shargon) [inline src/Plugins/RpcServer/RpcServer.cs:95]:
```suggestion
            // Always execute both checks, but both must evaluate to true
```

**Review by shargon — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

**Pull request merged**